auth-otp/Sources/App/Controllers/TOTPController.swift

```diff
@@ -0,0 +1,70 @@
+import Foundation
+import Hummingbird
+import HummingbirdAuth
+import OTP
+
+struct TOTPVerificationSession: Codable {
+    let secret: String
+    var verified: Bool
+}
+
+struct TOTPController<Users: UserRepository, Storage: PersistDriver> {
+    let users: Users
+    let storage: Storage
+
+    var endpoints: RouteCollection<AppRequestContext> {
+        let routes = RouteCollection(context: AppRequestContext.self)
+        routes
+            .group("users/totp", context: AppSessionRequestContext.self)
+            .addMiddleware {
+                SessionMiddleware(storage: self.storage)
+                SessionAuthenticator(users: self.users)
+            }
+            .post("start") { request, context -> EditedResponse<String> in
+                let otpSecret = UUID().uuidString
+                let otpSetupSession = UUID().uuidString
+                try await self.storage.create(key: otpSetupSession, value: TOTPVerificationSession(secret: otpSecret, verified: false), expires: .seconds(60 * 10))
+                return .init(
+                    headers: [.otpSession: otpSetupSession],
+                    response: TOTP(secret: otpSecret).createAuthenticatorURL(label: "auth-otp")
+                )
+            }
+            .post("verify/{code}") { request, context -> HTTPResponse.Status in
+                let code = try context.parameters.require("code", as: Int.self)
+                guard let otpSession = request.headers[.otpSession] else { return .badRequest }
+                guard var otpVerificationSession = try await self.storage.get(key: otpSession, as: TOTPVerificationSession.self) else { return .notFound }
+                let computedTOTP = TOTP(secret: otpVerificationSession.secret).compute(date: .now - 15.0)
+                let computedTOTP2 = TOTP(secret: otpVerificationSession.secret).compute(date: .now + 15.0)
```

Comment:
Very much a nit, but theoretically the distance between `.now` the first and second time can theoretically change enough to create an issue. I don't think we care for an example, but wanted to give you that info.